File: GreedyAlgo/job_sequencing_efficent.py
# Python program to find the maximum profit job sequence from a given array 
# of jobs with deadlines and profits 

# A class to represent a job 
class Job:
    def __init__(self, id, dead, profit):
        self.id = id    # Job Id 
        self.dead = dead    # Deadline of job 
        self.profit = profit    # Profit if job is over before or on deadline 

class Solution:
    @staticmethod
    def compare_by_profit(a, b):
        return a.profit > b.profit

    def job_scheduling(self, arr, n):

        total_profit = 0
        total_jobs = 0
        ans = [0, 0]
        reg_arr = [0] * (n+1)
        arr.sort(key=lambda x: x.profit, reverse=True)

        i = 0

        while i < n:

            if reg_arr[arr[i].dead] == 0:
                reg_arr[arr[i].dead] = arr[i].profit
                print(f"profit:{arr[i].profit} deadline:{arr[i].dead}")
            else:
                j = arr[i].dead - 1
               
                while j > 0:
                    
                    if reg_arr[j] == 0:
                        reg_arr[j] = arr[i].profit
                        break
                        print(f"profit:{arr[i].profit} deadline:{arr[i].dead}")

                    j -= 1

            i += 1

        for i in range(n+1):
            if reg_arr[i] > 0:
                total_profit += reg_arr[i]
                total_jobs += 1

        ans[0] = total_jobs
        ans[1] = total_profit
        return ans

# Driver program to test methods 
if __name__ == "__main__":
    n = 5
    arr = [Job(1, 3, 16), Job(2, 5, 263), Job(3, 3, 449), Job(4, 1, 310),Job(5,5,68)]

    ob = Solution()
    result_pair = ob.job_scheduling(arr, n)
    for p in result_pair:
        print(p)

File: GreedyAlgo/job_sequencing_efficent.cpp
#include <iostream>
#include <vector>
#include <unistd.h>

using namespace std;

//g++ -std=c++11 -O2 -Wall exp.cpp -o exp
//{ Driver Code Starts
// Program to find the maximum profit job sequence from a given array 
// of jobs with deadlines and profits 

// A structure to represent a job 
struct Job 
{ 
    int id;	 // Job Id 
    int dead; // Deadline of job 
    int profit; // Profit if job is over before or on deadline 
}; 



class Solution 
{

    public:
    //Function to find the maximum profit and the number of jobs done.
    bool static compareByProfit(Job& a, Job& b)
    {
        return a.profit>b.profit;
    }

    vector<int> JobScheduling(Job arr[], int n) 
 { 
      cout<<"hello";
        int totalProfit=0;
        int totalJob=0;
        vector<int> ans;
        vector<int> regArr(10,0);
        sort(arr,arr+n,compareByProfit);

   
        int i=0;
  cout<<"hello";
        while(i<n)
        {
              cout<<"hello";
            if(regArr[arr[i].dead]==0)
            {
                regArr[arr[i].dead]=arr[i].profit;

            }
            else
            {
                int j=arr[i].dead-1;
                cout<<"hello";
            

                while(j>0)
                {
                    if(regArr[j]==0)
                    {
                        regArr[j]=arr[i].profit;
                    }
                    j--;
                    
                }
                

                
            }
            i++;
         
        }

        for(int i=0;i<n;i++)
        {
            if(regArr[i]>0)
            {
                totalProfit+=regArr[i];
                totalJob+=1;

            }
            
        }
        ans[0]=totalJob;
        ans[1]=totalProfit;
        cout<<totalJob<<" :"<<totalProfit;
       return ans;
    } 
};

//{ Driver Code Starts.
// Driver program to test methods 
int main() 
{ 
      int n = 4;
   Job arr[] = {{1,4,20},{2,1,10},{3,2,40},{4,2,30}};

   Solution ob;
   vector<int> pair=ob.JobScheduling(arr,n);

	return 0; 
}






